src/report_reader.cpp

```diff
@@ -284,24 +284,63 @@ std::vector<NodeID> ReportReader<T>::Population::getNodeIds() const {
 }
 
 template <typename T>
-Selection::Values ReportReader<T>::Population::node_ids_from_selection(
+typename ReportReader<T>::Population::ElementIdsData ReportReader<T>::Population::getElementIds(
     const nonstd::optional<Selection>& selection) const {
-    Selection::Values node_ids;
-
-    if (!selection) {  // Take all nodes in this case
-        node_ids.reserve(nodes_pointers_.size());
-        std::transform(nodes_pointers_.begin(),
-                       nodes_pointers_.end(),
-                       std::back_inserter(node_ids),
-                       [](const std::pair<NodeID, Range>& node_pointer) {
-                           return node_pointer.first;
-                       });
-    } else if (selection->empty()) {
-        return {};
-    } else {
-        node_ids = selection->flatten();
+    ElementIdsData result;
+    result.range = {std::numeric_limits<uint64_t>::max(), std::numeric_limits<uint64_t>::min()};
+    size_t eids_count = 0;
+
+    // Helper function to update the min / max values and count
+    const auto& update_range = [&result, &eids_count](const Selection::Range& range_new) {
+        result.range.first = std::min(result.range.first, range_new.first);
+        result.range.second = std::max(result.range.second, range_new.second);
+        eids_count += (range_new.second - range_new.first);
+    };
+
+    // Take all nodes if no selection is provided
+    if (!selection) {
+        result.node_pointers = node_pointers_;
+
+        for (const auto& node_pointer : result.node_pointers) {
+            update_range(node_pointer.second);
+        }
+    } else if (!selection->empty()) {
+        const auto& node_ids = selection->flatten();
+
+        for (const auto& node_id : node_ids) {
+            const auto it = node_pointers_.find(node_id);
+            if (it != node_pointers_.end()) {
+                result.node_pointers.emplace(*it);
+                update_range(it->second);
+            }
+        }
+    }
+
+    // Extract the ElementIDs from the GIDs
+    if (!result.node_pointers.empty()) {
+        const auto min = result.range.first;
+        const auto max = result.range.second;
+
+        std::vector<ElementID> element_ids;
+        auto dataset_elem_ids = pop_group_.getGroup("mapping").getDataSet("element_ids");
+        dataset_elem_ids.select({min}, {max - min}).read(element_ids);
+
+        result.ids.reserve(eids_count);
+
+        for (const auto& node_pointer : result.node_pointers) {
+            const auto node_id = node_pointer.first;
+            const auto& range = node_pointer.second;
+
+            for (auto i = (range.first - min); i < (range.second - min); i++) {
+                result.ids.emplace_back(make_key<T>(node_id, element_ids[i]));
+            }
+        }
+
+        // Reduces overhead of next IOps by clearing-up internal buffers of HDF5
```

Comment:
@jorblancoa and I replicated the issue on a separate example using the C library of HDF5. It seems to happen when you ask for a large hyperslab in a dataset and then move to another dataset in the same file where you also ask for another large range. For some reason, the first cold access takes a few seconds extra and we observed that fooling HDF5 does the trick. The old implementation did not show the issue because it was conducting small ~300-value accesses.

The plan to report this issue to the HDFGroup, [just like we did when we reported the ASCII bug that we thought it was our fault in `libsonata`](https://github.com/HDFGroup/hdf5/issues/544), if you remember. For now, let's keep it, please.

Reply:
> For now, let's keep it, please.

Yeah, for sure!

It's such an odd thing, but I also tried it w/ the 'old' implementation, and saw a bit of speedup (not as big as this PR, obv, but still non-zero).

Can you add the explanation you just gave as a comment (`It seems to happen...`)?  It would also be helpful to add you found it, so we can test it again in the future, and potentially drop it if it's fixed by upstream HDF5.

Reply:
Yes, I agree. I just updated the code with further clarifications.